src/expr/src/linear.rs

```diff
@@ -889,23 +889,40 @@ impl MapFilterProject {
     /// );
     /// ```
     pub fn optimize(&mut self) {
-        // Optimization memoizes individual `ScalarExpr` expressions that
-        // are sure to be evaluated, canonicalizes references to the first
-        // occurrence of each, inlines expressions that have a reference
-        // count of one, and then removes any expressions that are not
-        // referenced.
-        self.memoize_expressions();
-        self.predicates.sort();
-        self.predicates.dedup();
-        self.inline_expressions();
-        self.remove_undemanded();
-
-        // Re-build `self` from parts to restore evaluation order invariants.
-        let (map, filter, project) = self.as_map_filter_project();
-        *self = Self::new(self.input_arity)
-            .map(map)
-            .filter(filter)
-            .project(project);
+        // Track sizes and iterate as long as they decrease.
+        let mut prev_size = usize::max_value();
+        let mut self_size = self.size();
+
+        // Continue as long as strict improvements occur.
+        while self_size < prev_size {
+            // Lock in current size.
+            prev_size = self_size;
+            // Optimization memoizes individual `ScalarExpr` expressions that
+            // are sure to be evaluated, canonicalizes references to the first
+            // occurrence of each, inlines expressions that have a reference
+            // count of one, and then removes any expressions that are not
+            // referenced.
+            self.memoize_expressions();
+            self.predicates.sort();
+            self.predicates.dedup();
+            self.inline_expressions();
+            self.remove_undemanded();
+
+            // Re-build `self` from parts to restore evaluation order invariants.
+            let (map, filter, project) = self.as_map_filter_project();
+            *self = Self::new(self.input_arity)
+                .map(map)
+                .filter(filter)
+                .project(project);
+
+            self_size = self.size();
+        }
+    }
+
+    /// Accumulate expression sizes across all expressions.
+    pub fn size(&self) -> usize {
+        self.expressions.iter().map(|e| e.size()).sum::<usize>()
+            + self.predicates.iter().map(|(_, e)| e.size()).sum::<usize>()
```

Comment:
Is it OK to call `e.size` repeatedly? This could be costly for large expression trees.

Reply:
Ah, it involves a traversal which we do all over the place elsewhere (e.g. comparing two expressions for equality). We are meant to call this only once for each pass through all optimizations (and repeat if it goes down).